### PR TITLE
add BOM module

### PR DESCRIPTION
**Pull request by mbenson**

### CODE REVIEW

**Comment by Hakky54** [comment]
Hi Matt, thank you this PR.

It seems like you are using multiple dependencies of this project and want to easily manage all of the versions right?
So what is your opinion regarding the transitive dependencies of each module, do they also need to be included in the bom, such as slf4j, bouncy castle, netty, jetty etc?

**Comment by Hakky54** [comment]
I am also not sure what approuch would be the best as I never have done something with either defining a bom or using it. I know it is useful and a great way to manage the dependencies. I need to do some testing to find out whether we also need to specify the transitive dependencies. It would be also usefull to find out what your test results are :)

**Comment by mbenson** [comment]
TBH, at this time the `pem` module, which as you know will bring in the "core" module transitively, is sufficient for my needs. I will test the BOM locally however for what it's worth.

**Comment by mbenson** [comment]
Actually, in testing I am seeing that because the BOM in its current state began by extending the existing parent, it is (via inheritance) managing `test`-scoped dependencies which IMO is overstepping. So if you are amenable to the basic _idea_ of the BOM we should A. reach consensus (for whatever my opinion is worth; you're the BDFL here ;) ) on what it should manage, and (possibly/probably) B. refactor the existing module hierarchy to achieve the desired outcome.

For my part, I believe that:
 * the BOM should manage _only_ sslcontext-kickstart's _own_ artifacts.
 * the BOM should inherit from the parent (to conveniently keep version in sync)
 * since the existing modules also share inherited dependency management, a possible approach would be to create a separate `dependencies` module to hold the managed dependencies that currently live in the parent. I will experiment with this locally.

**Comment by mbenson** [comment]
This additional commit implements my last point above and has the predicted effects, namely: the transitive slf4j-api dependency is resolved as expected, yet my BOM consumer is able to manage that dependency back to the previous version if for some reason that were needed (or if you didn't feel like updating this library sometime and a consumer wanted to upgrade that dependency in their own `dependencyManagement` section).

**Comment by mbenson** [comment]
I just happened across an example of a well-known BOM that includes only its "own" modules: https://search.maven.org/artifact/org.junit/junit-bom/5.9.2/pom

**Comment by Hakky54** [comment]
I see that you encapsulated the dependencies which the library itself is using, by moving it to a separate pom. The Bill of Material won't inherit those dependencies and therefor it wont appear in the effective pom of the end-user. It is a clever solution. Using similar setup as well know libraries is also a good way to way to go.

I just also checked spring and they have the same setup for the BOM, they only specify their own artifacts.
https://mvnrepository.com/artifact/org.springframework/spring-framework-bom/6.0.5
https://mvnrepository.com/artifact/org.junit/junit-bom/5.9.2

I checked their source code, for example this one: https://github.com/spring-projects/spring-framework/blob/0d10d4beeee3f33f69e4b1d84513e30ea08ccf66/spring-core/spring-core.gradle#L58

The dependency version is managed here in a separate configuration file: https://github.com/spring-projects/spring-framework/blob/0d10d4beeee3f33f69e4b1d84513e30ea08ccf66/framework-platform/framework-platform.gradle#L24

this looks similar to the pom you have created `sslcontext-kickstart-dependencies`

But the framework-platform is being applied to the spring bom, see here: https://github.com/spring-projects/spring-framework/blob/0d10d4beeee3f33f69e4b1d84513e30ea08ccf66/framework-bom/framework-bom.gradle#L3
 so it still does not really makes sense.

I will grab your your changes and try it locally, I will ping you when I have more insight about the behaviour of the current setup.

**Comment by Hakky54** [comment]
By the way which all dependencies are you using from this repository?

**Comment by mbenson** [comment]
For the moment, the PEM support (and core) were all I needed.

**Comment by Hakky54** [comment]
So in your case you only depend on `sslcontext-kickstart-for-pem` which will also provide you the core library as it has a transitive dependency. The other modules have also a transitive dependency on the core module so the end user only needs to specify 1 dependency such as your case. Although I like the idea of having a bom, but I don't think it is needed for now as the library is compact and having little submodules. What do you think?

**Comment by mbenson** [comment]
It's not needed for my purposes just at the moment, but it just struck me
that were I using more than one "leaf" module I would want the BOM. $0.02

**Comment by Hakky54** [comment]
Hmm, yes if someone uses more than 1 leaf module it would be handy. The project has already 5 leaf modules and might get more in the future. Yes lets add the support for a bom. I tried your branch locally and I have some remarks.

I found the explanation of Apache regarding the Bill of Materials useful and I would prefer to have their suggestion in parent-child relation. See here for the details: https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#bill-of-materials-bom-poms

So basically the bom is the parent, but we just call it bom which you already did. The bom will be inherrited by the parent and we can keep the name parent as it will contain all of the properties, plugins, dependency management so no need for another module having the suffix dependencies. I would prefer to have not too much inheritance. Al other leaf modules will inherit from the parent pom. It is almost similar to your suggestion but slightly different.

**Comment by Hakky54** [comment]
It seems like a kind of separate leaf module is needed such as the one you created or else maven will complain. I am just trying it out locally and checking whether I can have a working pom structure

File: pom.xml [hunk 1]
```diff
@@ -12,6 +12,8 @@
     <url>https://sslcontext-kickstart.com/</url>
 
     <modules>
+        <module>sslcontext-kickstart-bom</module>
+        <module>sslcontext-kickstart-dependencies</module>
```
Comment (Hakky54) [inline pom.xml:16]: This module can be dropped as I don't think it is required

File: pom.xml [hunk 2]
```diff
@@ -82,53 +79,6 @@
...
-                <scope>test</scope>
-            </dependency>
-            <dependency>
-                <groupId>io.github.hakky54</groupId>
-                <artifactId>logcaptor</artifactId>
-                <version>${version.logcaptor}</version>
-                <scope>test</scope>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
```
Comment (Hakky54) [inline pom.xml:130]: All of these changes can be reverted except:
```xml
            <dependency>
                <groupId>io.github.hakky54</groupId>
                <artifactId>sslcontext-kickstart</artifactId>
                <version>${project.version}</version>
            </dependency>
```

That one is not required

File: pom.xml [hunk 3]
```diff
@@ -82,53 +79,6 @@
...
-
```
Comment (Hakky54) [inline pom.xml:131]: All of these changes can be reverted except:
```xml
            <dependency>
                <groupId>io.github.hakky54</groupId>
                <artifactId>sslcontext-kickstart</artifactId>
                <version>${project.version}</version>
            </dependency>
```

That one is not required

File: sslcontext-kickstart-bom/pom.xml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.github.hakky54</groupId>
+        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <version>7.4.10-SNAPSHOT</version>
+    </parent>
+    <artifactId>sslcontext-kickstart-bom</artifactId>
+    <name>sslcontext-kickstart-bom</name>
```
Comment (Hakky54) [inline sslcontext-kickstart-bom/pom.xml:10]: I would prefer the name slighty different, what do think of `SSLContext Kickstart (Bill of Materials)` and also for description?

File: pom.xml [hunk 5]
```diff
@@ -12,6 +12,8 @@
     <url>https://sslcontext-kickstart.com/</url>
```
Comment (Hakky54) [inline pom.xml:13]: I think the following statement is needed:
```xml
    <parent>
        <groupId>io.github.hakky54</groupId>
        <artifactId>sslcontext-kickstart-bom</artifactId>
        <version>7.4.10-SNAPSHOT</version>
        <relativePath>sslcontext-kickstart-bom</relativePath>
    </parent>
```

File: sslcontext-kickstart-for-apache4/pom.xml [hunk 6]
```diff
@@ -4,8 +4,9 @@
 
     <parent>
         <groupId>io.github.hakky54</groupId>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-apache4/pom.xml:7]: this can be reverted

File: sslcontext-kickstart-for-apache4/pom.xml [hunk 7]
```diff
@@ -4,8 +4,9 @@
...
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-apache4/pom.xml:9]: this can be reverted

File: sslcontext-kickstart-for-apache5/pom.xml [hunk 8]
```diff
@@ -3,9 +3,10 @@
     <modelVersion>4.0.0</modelVersion>
 
     <parent>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-apache5/pom.xml:6]: this can be reverted

File: sslcontext-kickstart-for-apache5/pom.xml [hunk 9]
```diff
@@ -3,9 +3,10 @@
...
         <groupId>io.github.hakky54</groupId>
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-apache5/pom.xml:9]: this can be reverted

File: sslcontext-kickstart-for-jetty/pom.xml [hunk 10]
```diff
@@ -4,8 +4,9 @@
 
     <parent>
         <groupId>io.github.hakky54</groupId>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-jetty/pom.xml:7]: this can be reverted

File: sslcontext-kickstart-for-jetty/pom.xml [hunk 11]
```diff
@@ -4,8 +4,9 @@
...
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-jetty/pom.xml:9]: this can be reverted

File: sslcontext-kickstart-for-netty/pom.xml [hunk 12]
```diff
@@ -4,8 +4,9 @@
 
     <parent>
         <groupId>io.github.hakky54</groupId>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-netty/pom.xml:7]: this can be reverted

File: sslcontext-kickstart-for-netty/pom.xml [hunk 13]
```diff
@@ -4,8 +4,9 @@
...
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-netty/pom.xml:9]: this can be reverted

File: sslcontext-kickstart-for-pem/pom.xml [hunk 14]
```diff
@@ -4,8 +4,9 @@
 
     <parent>
         <groupId>io.github.hakky54</groupId>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-pem/pom.xml:7]: this can be reverted

File: sslcontext-kickstart-for-pem/pom.xml [hunk 15]
```diff
@@ -4,8 +4,9 @@
...
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart-for-pem/pom.xml:9]: this can be reverted

File: sslcontext-kickstart/pom.xml [hunk 16]
```diff
@@ -4,8 +4,9 @@
 
     <parent>
         <groupId>io.github.hakky54</groupId>
-        <artifactId>sslcontext-kickstart-parent</artifactId>
+        <artifactId>sslcontext-kickstart-dependencies</artifactId>
```
Comment (Hakky54) [inline sslcontext-kickstart/pom.xml:7]: this can be reverted

File: sslcontext-kickstart/pom.xml [hunk 17]
```diff
@@ -4,8 +4,9 @@
...
         <version>7.4.10-SNAPSHOT</version>
+        <relativePath>../sslcontext-kickstart-dependencies</relativePath>
```
Comment (Hakky54) [inline sslcontext-kickstart/pom.xml:9]: this can be reverted

File: sslcontext-kickstart-dependencies/pom.xml [hunk 18]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Hakky54) [inline sslcontext-kickstart-dependencies/pom.xml:1]: This pom and leaf module can be deleted

**Comment by Hakky54** [comment]
I dropped couple of remarks. would it be possible for you to do some rework?

**Comment by mbenson** [comment]
RE rework: I can implement your requested changes, although rather than reverting all the changes you mentioned I think it'd be more efficient for me to simply drop the second of my commits and start there (which is one reason I didn't squash that commit to begin with). My read of the Maven BOM documentation is that their example is structured a little differently than yours (parent is physically nested into bom and presumably other modules would refer by relative path _../parent_) but if the structure you describe is working that's fine with me. I will also add the change you requested to name/description of the BOM module.

**Comment by mbenson** [comment]
I do have a question/clarification for you regarding your experimentation: you _are_ saying you kept the parent module at the _physical_ root with its `parent` element pointing to the BOM module by `relativePath>sslcontext-kickstart-bom`? Which module contains the "root" `version`? I am concerned/confused as to how e.g. the `versions:set` goal will work, whether an `install` from the root will implicitly install the BOM even though it's not (seemingly) explicitly built... etc.

Or are you saying that the BOM would become the root and the parent would move to be physically nested below as in the example documentation?

**Comment by mbenson** [comment]
As I feared, the structure where parent both physically nests and extends BOM is problematic w/ `versions:set` and suffers from the issue that `mvn install` executed from the project root fails to install the BOM. You can of course `cd` into the BOM directory and install it as well, but clearly that is less convenient.

**Comment by Hakky54** [comment]
what you can try is make the change in the bom module pom file you can specify a property like this:

```xml
<properties>
    <revision>7.4.10-SNAPSHOT</revision>
</properties>
```

And change all versions within every other pom like the snippet below:
```xml
<version>${revision}</version>
```

Can you maybe retry with the setup above?

**Comment by mbenson** [comment]
Having done that and added `<module>..</module>` to the BOM, everything hangs together pretty well with the caveat that you have to `cd` into the BOM's directory to `mvn install` the whole set of modules. So it works although it's maybe fairly counterintuitive. What I could do is commit this structure to a separate branch and also prepare a variant emulating the documentation you referenced so that we could more easily contrast the three options.

**Comment by Hakky54** [comment]
Can you push it? I am curious and I would like to try it out 😄

**Comment by mbenson** [comment]
To be clear, what I just added here was the non-controversial change to BOM name/desc. I have pushed the current state to https://github.com/mbenson/sslcontext-kickstart/tree/bom-under-pom and as I described in the last commit message, I consider this to be described as "derive all modules from BOM with actual project root as subdirectory of physical root".

**Comment by Hakky54** [comment]
Yes, I see the issue you mentioned. The bom is not being build. I just found a hacky way to get it working.

If you remove the below snippet from the bom:
```xml
<modules>
    <module>..</module>
</modules>
```

And add the BOM module to the parent pom modules. So something like this:
```xml
<modules>
    <module>sslcontext-kickstart</module>
    <module>sslcontext-kickstart-for-netty</module>
    <module>sslcontext-kickstart-for-jetty</module>
    <module>sslcontext-kickstart-for-apache4</module>
    <module>sslcontext-kickstart-for-apache5</module>
    <module>sslcontext-kickstart-for-pem</module>
    <module>sslcontext-kickstart-bom</module>
</modules>
```

I think this should do the trick. Can you retry on your side? I am curious whether it will also work for you.

**Comment by mbenson** [comment]
When I do this, I get:
```
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project io.github.hakky54:sslcontext-kickstart-parent:7.4.10-SNAPSHOT (~/dev/oss/sslcontext-kickstart/pom.xml) has 1 error
[ERROR]     Child module ~/dev/oss/sslcontext-kickstart/sslcontext-kickstart-bom/pom.xml of ~/dev/oss/sslcontext-kickstart/pom.xml forms aggregation cycle ~/dev/oss/sslcontext-kickstart/sslcontext-kickstart-bom/pom.xml -> ~/dev/oss/sslcontext-kickstart/pom.xml -> ~/dev/oss/sslcontext-kickstart/sslcontext-kickstart-bom/pom.xml
```

**Comment by Hakky54** [comment]
Hmm, strange. Can you push your changes and change the source branch in this PR? So we can start running the build pipeline on github. And it is weird that I am not getting that error locally

**Comment by mbenson** [comment]
See my edit to my previous comment. That is working for me, after all. Even `mvn versions:set` works if you explicitly specify `-pl sslcontext-kickstart-bom`.

**Comment by Hakky54** [comment]
Yess, we are making very good progress. Any idea whether you can change the source branch with this pull request? Or else you can pull your changes from the other branch and drop your initial changes from the initial branch

**Comment by mbenson** [comment]
Actually it turns out that `versions:set` as above _technically_ works, but it wipes out those `${revision}` properties. I could experiment with whether we could abandon them now that we figured out the module structure.

Certainly I can change the commits on this branch (but I'm still saving the old ones locally until we are confident with an approach). I will push the latest state.

**Comment by Hakky54** [comment]
I am never using `mvn versions:set`, when and why should I use that? I use maven release plugin to bump the version and that works with the current change what you have 😄 

Thank you by the way for convincing me to support a bom for this project, A appreciate your convincing powers. I think the end results looks good and ready to merge, what do you think? Are you happy with the result?

**Comment by mbenson** [comment]
If you are satisfied that `maven-release-plugin` works fine with this structure we can go with it. Are you interested to try reverting the `${revision}` bit or would you prefer to keep it?

**Comment by mbenson** [comment]
(you may also have noticed that I left BOM as the first defined module to reflect its "higher" status)

**Comment by Hakky54** [comment]
> (you may also have noticed that I left BOM as the first defined module to reflect its "higher" status)

That's ok for me.

> Are you interested to try reverting the ${revision} bit or would you prefer to keep it?

I am fine of keeping the revision property as is and the usage within the bom and parent

**Comment by mbenson** [comment]
I happened to poke around after I noticed you'd merged and saw you'd done the release. Excited to pull that in tomorrow or so whenever Sonatype syncs it out there :)

**Comment by Hakky54** [comment]
There is an issue with version 7.4.10 which I had released an hour ago. Somehow it didn't publish the bom. I applied a fix and did the deployment again and now it is available from version 7.4.11 onwards.
